### PR TITLE
IterationMapper

### PR DESCRIPTION
**Pull request by divyanshiGupta**

### CODE REVIEW

File: src/app/effects/iteration.effects.ts [hunk 1]
```diff
@@ -12,9 +14,14 @@ export class IterationEffects {
   }
 
   @Effect() getIterations$ : Observable<Action> = this.actions$
-    .ofType(IterationActions.GET)
+    .ofType<IterationActions.Get>(IterationActions.GET)
     .switchMap(action => {
       return this.iterationService.getIterations()
+           .do(iterations=> {
```
Comment (sanbornsen) [inline src/app/effects/iteration.effects.ts]: How much space is it for indentation?

File: src/app/effects/iteration.effects.ts [hunk 2]
```diff
@@ -12,9 +14,14 @@ export class IterationEffects {
   }
 
   @Effect() getIterations$ : Observable<Action> = this.actions$
-    .ofType(IterationActions.GET)
+    .ofType<IterationActions.Get>(IterationActions.GET)
     .switchMap(action => {
       return this.iterationService.getIterations()
+           .do(iterations=> { 
+                              let iterationMapper: IterationMapper;
```
Comment (sanbornsen) [inline src/app/effects/iteration.effects.ts]: What is the use of this variable?
```:TypeName``` is used to define a type not creating an instance of a class. Why not use ```new IterationMapper()``` instead ?

File: src/app/models/iteration.model.ts [hunk 3]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
   startAt: string;
   endAt: string;
   description: string;
+  state: string;
+  links: string;
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: As of now we just need one link, that is the self link, So the key should be ```link```.

File: src/app/models/iteration.model.ts [hunk 4]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+  workItemCount: number;
+  type: string;
+
+}
+
+export class IterationMapper {
+iterationModel: IterationModel;
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Please follow proper indentation.

File: src/app/models/iteration.model.ts [hunk 5]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+iterationsUI: IterationUI[];
+
+
+IterationModeltoIterationUI(iterations: IterationModel[]): IterationUI[] {
+  let i;
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: What is this extra new line for?

File: src/app/models/iteration.model.ts [hunk 6]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+  for(i=0;i<iterations.length;i=i+1)
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Why can't ```i``` be initialized inside the loop?
Maintain spaces inside the for statement (follow the linters)
```for(let i = 0; i < iterations.length; i++)```

File: src/app/models/iteration.model.ts [hunk 7]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+  state: string;
+  links: string;
+  workItemCount: number;
+  type: string;
+
+}
+
+export class IterationMapper {
+iterationModel: IterationModel;
+iterationsUI: IterationUI[];
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: There is no interface as ```iterationsUI```, it should be singular.

File: src/app/models/iteration.model.ts [hunk 8]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+
+
+IterationModeltoIterationUI(iterations: IterationModel[]): IterationUI[] {
+  let i;
+  for(i=0;i<iterations.length;i=i+1)
+  {
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: What is this extra blank lines for?

File: src/app/models/iteration.model.ts [hunk 9]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+  type: string;
+
+}
+
+export class IterationMapper {
+iterationModel: IterationModel;
+iterationsUI: IterationUI[];
+
+
+IterationModeltoIterationUI(iterations: IterationModel[]): IterationUI[] {
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: This function name should be same for all the attributes. Let's keep it ```toUIModel```. There should be an interface for mapper classes.

File: src/app/models/iteration.model.ts [hunk 10]
```diff
@@ -64,4 +64,81 @@ export interface IterationUI extends modelUI {
...
+    
+    this.iterationsUI[i] = this._utilMapperUIModel(iterations[i]);
+    
+  }
+
+  return this.iterationsUI;
+
 }
+
+IterationUItoIterationModel(iteration: IterationUI): IterationModel {
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Similarly this function should also be named, ```toServiceModel```

File: src/app/models/iteration.model.ts (hunk 9 above)
Comment (sanbornsen) [reply src/app/models/iteration.model.ts]: This function should take one iteration and return one uteration.

**Review by sanbornsen — CHANGES_REQUESTED** [review]

File: src/app/models/iteration.model.ts (hunk 9 above)
Comment (divyanshiGupta) [reply src/app/models/iteration.model.ts]: This is done by _utilMapperUIModel as the above function calls this utility function with single iteration and the utility function returns a single iteration after mapping it to UI model which is then added to array of the UI model.

File: src/app/effects/iteration.effects.ts [hunk 11]
```diff
@@ -3,7 +3,9 @@ import { Actions, Effect } from "@ngrx/effects";
 import { Observable } from "rxjs";
 import * as IterationActions from ".././actions/iteration.actions";
 import { IterationService } from '.././services/iteration.service';
-import { Action } from '@ngrx/store';
+import { IterationMapper } from '.././models/iteration.model';
```
Comment (sanbornsen) [inline src/app/effects/iteration.effects.ts]: Why not just ```./../``` ? What is the use of ```/./``` ?

File: src/app/effects/iteration.effects.ts [hunk 12]
```diff
@@ -12,10 +14,14 @@ export class IterationEffects {
...
 
   @Effect() getIterations$ : Observable<Action> = this.actions$
-    .ofType(IterationActions.GET)
+    .ofType<IterationActions.Get>(IterationActions.GET)
     .switchMap(action => {
       return this.iterationService.getIterations()
-           .map(iterations => (new IterationActions.GetSuccess(iterations)))
-           .catch(() => Observable.of(new IterationActions.GetError()))
+       .do(iterations=> { 
+                          return new IterationMapper().toUIModel(iterations);
```
Comment (sanbornsen) [inline src/app/effects/iteration.effects.ts]: I still see indentation problem here.

File: src/app/models/iteration.model.ts [hunk 13]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+
+export class IterationMapper {
+
+  iterationModel: IterationModel;
+  iterationUI: IterationUI[];
+
+
+  toUIModel(iterations: IterationModel[]): IterationUI[] {
+
+   for(let i=0; i<iterations.length; i=i+1)
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: There should be spaces between operators. ```i = 0```, ```i < iterations.length``` etc.

File: src/app/models/iteration.model.ts [hunk 14]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
   startAt: string;
   endAt: string;
   description: string;
+  state: string;
+  link: string;
+  workItemCount: number;
+  type: string;
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Remove this empty line.

File: src/app/models/iteration.model.ts [hunk 15]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+export class IterationMapper {
+
+  iterationModel: IterationModel;
+  iterationUI: IterationUI[];
+
+
+  toUIModel(iterations: IterationModel[]): IterationUI[] {
+
+   for(let i=0; i<iterations.length; i=i+1)
+   {
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Curly braces start from the line above to maintain consistency.

File: src/app/models/iteration.model.ts [hunk 16]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+     this.iterationUI[i] = this._utilMapperUIModel(iterations[i]);
+   }
+   
+   return this.iterationUI;
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Remove empty line.

File: src/app/models/iteration.model.ts [hunk 17]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+  iterationUI: IterationUI[];
+
+
+  toUIModel(iterations: IterationModel[]): IterationUI[] {
+
+   for(let i=0; i<iterations.length; i=i+1)
+   {
+     this.iterationUI[i] = this._utilMapperUIModel(iterations[i]);
+   }
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Remove empty line

File: src/app/models/iteration.model.ts [hunk 18]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+   {
+     this.iterationUI[i] = this._utilMapperUIModel(iterations[i]);
+   }
+   
+   return this.iterationUI;
+
+  }
+
+  toServiceModel(iteration: IterationUI): IterationModel {
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Remove empty lines.

File: src/app/models/iteration.model.ts [hunk 19]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+     this.iterationUI[i] = this._utilMapperUIModel(iterations[i]);
+   }
+   
+   return this.iterationUI;
+
+  }
+
+  toServiceModel(iteration: IterationUI): IterationModel {
+  
+   this.iterationModel = this._utilMapperServiceModel(iteration);
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Improper indentation.

File: src/app/models/iteration.model.ts [hunk 20]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+   return this.iterationModel;
+
+  }
+
+
+  _utilMapperServiceModel(iteration: IterationUI): IterationModel {
+   
+   let iterationModel: IterationModel;
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Improper indentation.

File: src/app/models/iteration.model.ts [hunk 21]
```diff
@@ -64,4 +64,80 @@ export interface IterationUI extends modelUI {
...
+   iterationModel.id = iteration.id;
+   iterationModel.links.self = iteration.link;
+   iterationModel.relationships.workitems.meta.total = iteration.workItemCount;
+   iterationModel.type = iteration.type;
+
+   return iterationModel;
+  }
+
+  _utilMapperUIModel(iterationModel: IterationModel): IterationUI {
+
```
Comment (sanbornsen) [inline src/app/models/iteration.model.ts]: Remove empty line

**Review by sanbornsen — CHANGES_REQUESTED** [review]

**Comment by sanbornsen** [comment]
@divyanshiGupta can you write a mapper spec file with some test cases? We just some unit test to test this functions out.

**Comment by divyanshiGupta** [comment]
Ok. I will do that

**Pull request closed without merging**